Added internationalized screen reader helpers to EuiMark.

1Copenut opened the pull request:
### Summary

This issue adds screen reader helper text to denote the beginning and end of a highlighted passage of text. The screen reader text uses the `i18n` helper so they are translated properly.

This PR closes #3072. 

### Checklist

- [ ] Checked in both **light and dark** modes
- [ ] Checked in **mobile**
- [ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [ ] ~~Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**~~
- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples
- [ ] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
- [ ] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [ ] ~~Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart~~
- [ ] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

thompsongl commented on upcoming_changelogs/5739.md:
```diff
@@ -0,0 +1 @@
+- Added internationalized screen reader helpers to `EuiMark`. [(#5739)](https://github.com/elastic/eui/pull/5739)
```
```suggestion
- Added internationalized screen reader helpers to `EuiMark`
```

Don't need to manually add the link anymore 🎊 The script will do it automatically now.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5739/

thompsongl commented:
Raising this for discussion because I think I looked into this a couple years ago:

Now that we're committed to CSS-in-JS, we could use this approach: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/mark#accessibility_concerns

I think i18n may have been a blocker previously. Almost certain that there are tradeoffs here that I'm not aware of that perhaps makes your approach better.

cchaos commented:
Thinking through how `EuiHighlight` is generally used, which is to show where strings are matching... Are we concerned about this repetition when we get into places like EuiComboBox that might have hundreds of matching items and as the user arrows through, they constantly have to hear this "highlight start/end" text on each option?
<img width="472" alt="Screen Shot 2022-03-21 at 23 13 35 PM" src="https://user-images.githubusercontent.com/549577/159400836-4b6550e5-8d0b-45a6-9688-2a571b00ec9e.png">

thompsongl commented:
To clarify my previous comment:


Add this to the existing `css` block in `mark.styles.ts`:

```css
&:before,
&:after {
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
```

In `mark.tsx`:

```tsx

...

const highlightStart = useEuiI18n('euiMark.highlightStart', 'highlight start');
const highlightEnd = useEuiI18n('euiMark.highlightEnd', 'highlight end');

const pseudoStyles = css`
&:before {
  content: " [${highlightStart}] ";
}

&:after {
  content: " [${highlightEnd}] ";
}
`;

...


<mark css={[styles, pseudoStyles]} className={classes} {...rest}>
  {children}
</mark>
```

1Copenut commented:
I've refactored the CSS to use Emotion. Thanks @thompsongl for the code snippet.

@cchaos I tested with VoiceOver, and it does repeat the `:before` and `:after` blocks in every combo box option listing. It was verbose, but not overwhelming. I'll ask Zuhair his opinion tomorrow.

cchaos commented on src/components/mark/mark.tsx:
```diff
@@ -27,9 +29,23 @@ export const EuiMark: FunctionComponent<EuiMarkProps> = ({
   const useTheme = useEuiTheme();
   const styles = euiMarkStyles(useTheme);
```
Quick thought around compiled styles and the requirement for pseudo content...

Can we actually pass the i18n'd string into the `euiMarkStyles` function? So something like:

```suggestion
  const styles = euiMarkStyles({ theme: useTheme, highlightStartText: highlightStart, highlightEndText: highlightEnd });
```

cchaos commented on src/components/mark/mark.styles.ts:
```diff
@@ -22,5 +22,16 @@ export const euiMarkStyles = ({ euiTheme, colorMode }: UseEuiTheme) => {
     // Can't use 'inherit' because the text to background color contrast may not be sufficient
     color: ${euiTheme.colors.text};
+
+    &:before,
+    &:after {
+      clip-path: inset(100%);
+      cli: rect(1px, 1px, 1px, 1px);
```
So that way we can **conditionally** render these styles if the before/after content exists **and** pass the `content` through to this whole block.

I'm thinking about ways that consumers, for whatever reason, may want to turn off this SR text (they'd need a really good excuse 😉 ) and then they won't have all this extra pseudo styles for a pseudo element that doesn't exist.

1Copenut replied:
I'll take a run at it after dinner tonight, and if that comes out not so great, I'll defer to you @cchaos. These comments are solid and should get me started. Thank you!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5739/

1Copenut commented:
Ready for re-test. Thanks team!

cchaos commented on src/components/mark/mark.styles.ts:
```diff
@@ -22,5 +28,25 @@ export const euiMarkStyles = ({ euiTheme, colorMode }: UseEuiTheme) => {
+    // https://seanconnolly.dev/emotion-conditionals
+    ${hasHighlightText === true &&
+    `
+      &:before,
+      &:after {
+        content: ' [${highlightStart}] ';
+        clip-path: inset(100%);
+        clip: rect(1px, 1px, 1px, 1px);
+        height: 1px;
+        overflow: hidden;
+        position: absolute;
+        white-space: nowrap;
+        width: 1px;
+      }
+
+      &:after {
```
Nit picky: But I'd rather see the explicit `before` vs `after` text in their own. I find it easier than hunting for cascades.

```suggestion
      &:before,
      &:after {
        clip-path: inset(100%);
        clip: rect(1px, 1px, 1px, 1px);
        height: 1px;
        overflow: hidden;
        position: absolute;
        white-space: nowrap;
        width: 1px;
      }
      
      &:before {
        content: ' [${highlightStart}] ';        
      }

      &:after {
```

cchaos commented on src/components/mark/mark.tsx:
```diff
@@ -8,11 +8,17 @@
+     * Set to `false` to remove the CSS :before and :after
+     * screen reader helper text
+     */
+    hasHighlightText?: boolean;
```
Nit picky naming, but I'd rather have this be something named more screen-reader specific since I don't know what "highlight" here means in context of `EuiMark`
```suggestion
    hasScreenReaderHelpText?: boolean;
```

or something shorter... naming. hard.

cchaos commented on src/services/theme/hooks.tsx:
```diff
@@ -23,6 +22,9 @@ export interface UseEuiTheme<T extends {} = {}> {
   euiTheme: EuiThemeComputed<T>;
   colorMode: EuiThemeColorModeStandard;
   modifications: EuiThemeModifications<T>;
+  hasHighlightText?: boolean;
+  highlightStart?: string;
+  highlightEnd?: string;
```
DId you mean to add these to the global theme? It should really only be EuiMark specific.

cchaos commented on src/components/mark/mark.styles.ts:
```diff
@@ -9,7 +9,13 @@
+  hasHighlightText,
+  highlightStart,
+  highlightEnd,
+}: UseEuiTheme) => {
```
Oh I don't think you can use the type `UseEuiTheme` here because only the original 2 props were of this type. We should be extending this prop type not editing it globally.

thompsongl replied:
Yeah let's make it a second arg:

```
(
  { euiTheme, colorMode }: UseEuiTheme,
  hasHighlightText: boolean
)
```

or

```
(
  { euiTheme, colorMode }: UseEuiTheme,
  { hasHighlightText }: { hasHighlightText: boolean }
)
```

cchaos reviewed (changes requested):
The concept works! Just had some suggestions and found one issue with the TS

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5739/

1Copenut commented:
Ready for next iteration review. I added the `hasScreenReaderHelpText` prop to the EuiHighlight component and updated unit test to account for the default and override behavior.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5739/

1Copenut commented:
jenkins please retest

chandlerprall commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5739/

cchaos commented on src/components/highlight/highlight.test.tsx:
```diff
@@ -42,6 +42,20 @@ describe('EuiHighlight', () => {
       });
+
+      test('is rendered without CSS :before and :after text', () => {
+        const component = render(
```
🤔 I'm still not seeing the rendered styles in the snapshot for highlight specifically. Either we're really testing the output of EuiMark here, or just ensuring that the prop is passed down. If the latter, I'd just change the name of the test:

```tsx
test('hasScreenReaderHelpText can be false', () => {
```

cchaos replied:
Just renaming the test then should be fine

1Copenut replied:
Changed and new code ready for review.

cchaos reviewed (approved):
LGTM, but I'll leave it to @thompsongl to do the eng review.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5739/

1Copenut commented:
I talked with Zuhair today and affirmed this is working well in NVDA and JAWS. Shipping it. Thanks everyone.
